src/org/traccar/protocol/OpenGtsProtocolDecoder.java

```diff
@@ -0,0 +1,141 @@
+            .expression("([AV]),")               // validity
+            .number("(d+)(dd.d+),")              // latitude
+            .expression("([NS]),")
+            .number("(d+)(dd.d+),")              // longitude
+            .expression("([EW]),")
+            .number("(d+.d+),")                  // speed
+            .number("(d+.d+),")                 // course
+            .number("(dd)(dd)(dd),")             // date (ddmmyy)
+            .any()
+            .compile();
+
+    public OpenGtsProtocolDecoder(OpenGtsProtocol protocol) {
+        super(protocol);
+    }
+
+    @Override
+    protected Object decode(
+            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        HttpRequest request = (HttpRequest) msg;
+        QueryStringDecoder decoder = new QueryStringDecoder(request.getUri());
+        Map<String, List<String>> params = decoder.getParameters();
+        if (params.isEmpty()) {
+            decoder = new QueryStringDecoder(request.getContent().toString(StandardCharsets.US_ASCII), false);
+            params = decoder.getParameters();
+        }
+
+        Position position = new Position();
+        position.setProtocol(getProtocolName());
+        position.setValid(true);
+
+        for (Map.Entry<String, List<String>> entry : params.entrySet()) {
+            String value = entry.getValue().get(0);
+            switch (entry.getKey()) {
+                case "id":
+                    DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, value);
+                    if (deviceSession == null) {
+                        sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+                        return null;
+                    }
+                    position.setDeviceId(deviceSession.getDeviceId());
+                    break;
+                case "gprmc":
+                    Parser parser = new Parser(PATTERN, value);
+                    if (!parser.matches()) {
+                            sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+                        return null;
+                    }
+
+                    DateBuilder dateBuilder = new DateBuilder()
+                            .setTime(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
+
+                    position.setValid(parser.next().equals("A"));
+                    position.setLatitude(parser.nextCoordinate());
+                    position.setLongitude(parser.nextCoordinate());
+                    position.setSpeed(parser.nextDouble(0));
+                    position.setCourse(parser.nextDouble(0));
+
+                    dateBuilder.setDateReverse(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
+                    position.setTime(dateBuilder.getDate());
+                    break;
+                case "alt":
+                    position.setAltitude(Double.parseDouble(value));
+                    break;
+                case "batt":
```

Comment:
Are all parameters real? Please include only ones that OpenGTS protocol actually has.

Reply:
Yes, as you could see from test string there is actual altitude and battery. Depending on device battery reported zero or real value. Altitude seems real on all devices I have.

Reply:
OK.